src/controllers/eventsController.js

```diff
@@ -0,0 +1,135 @@
+const Group = require("../models/Group");
+const Event = require("../models/Event");
+const EventActivity = require("../models/EventActivity");
+
+// @desc    Endpoint for fetching all events
+// @routes  GET /api/v1/events
+//          GET /api/v1/events?host=true
+// @access  signed in users only
+const getAllEvents = async (req, res) => {
+  const { userID } = req.user;
+  const { host } = req.query;
+
+  const findCriteria = host ? { host: userID } : {};
+
+  const response = await Event.find(findCriteria).populate("activities");
+
+  return res.json({ count: response.length, events: response });
+};
+
+// @desc    Endpoint for fetching a single event
+// @route   GET /api/v1/events/:_id
+// @access  signed in users only
+const getEvent = async (req, res) => {
+  const { _id } = req.params;
+
+  const event = await Event.findOne({ _id })
+    .populate({
+      path: "activities",
+      select: "_id eventID activity type votes",
+    })
+    .populate({
+      path: "participants",
+      select: "_id name email",
+    })
+    .populate({
+      path: "host",
+      select: "_id name email",
+    });
+  if (!event) {
+    res.status(404);
+    throw new Error(`Event with ${_id} ID does not exist!`);
+  }
+
+  return res.json(event);
+};
+
+// @desc    Endpoint for creating an event
+// @route   POST /api/v1/events
+// @access  signed in users only
+const saveEvent = async (req, res) => {
+  let { groupID, name, description, eventDateTime, activities } = req.body;
+  const { userID } = req.user;
+
+  if (!groupID || !(await Group.findOne({ _id: groupID }))) {
+    res.status(400);
+    throw new Error(
+      "A valid 'groupID' value must be provided in order to link the event to its group!"
+    );
+  } else if (!name) {
+    res.status(400);
+    throw new Error("Event must have a name!");
+  } else if (!eventDateTime) {
+    res.status(400);
+    throw new Error("Event must have a date and time!");
+  } else if (activities) {
+    const isValidActivities =
+      Array.isArray(activities) &&
+      activities.every((activity) => {
+        return (
+          activity.hasOwnProperty("activity") && activity.hasOwnProperty("type")
+        );
+      });
+    if (!isValidActivities) {
+      res.status(400);
+      throw new Error(
+        "Invalid 'activities' format. An array of at least one activity object that must have 'name' and 'type' properties!"
+      );
+    }
+  }
+  const savedEvent = new Event({
+    groupID,
+    name,
+    host: userID,
+    eventDateTime,
+    description,
+  });
+  activities = activities.map((element) => {
+    return { eventID: savedEvent._id, ...element };
+  });
+  const savedEventActivities = await EventActivity.create(activities);
+  for (activity of savedEventActivities) {
+    savedEvent.activities.push(activity._id);
+  }
+  await savedEvent.save();
```

Comment:
Yes. Though my only decision for designing it this is because if we were to create an event activity's page on its own with many other detailed data fields in the future, referencing the event on the `EventActivity` is good idea. But you are correct, having one of the references does simplify things.

Reply:
This is another example of a [one-to-many](https://www.sqltutorial.net/sql-one-to-many-relationship.html) relationship (an `Event` can have many `EventActivities`, each `EventActivity` belongs to an `Event`), so I actually think the best choice would be to only reference the event on the `EventActivity`, and remove the array of `EventActivity` refs from the `Event`. We can query `EventActivities` using the event id when populating data for the `GET /events/:id` endpoint.

If you'd like, we can just create a task to refactor this and come back to it later (if there's time). As I mentioned to Marice, since we're getting close to demo time, we might decide to take on some ["tech debt"]([tech](https://asana.com/resources/technical-debt)) in order to prioritize shipping faster and getting more done. But it's up to y'all! 🙂